traits/trait_dict_object.py

```diff
@@ -57,18 +50,314 @@ def __init__(self, added=None, changed=None, removed=None):
+
+        Parameters
+        ----------
+        key : A hashable object.
+            The key to be deleted.
+
+        Raises
+        ------
+        KeyError
+            If the key is not found.
+        """
+
+        removed = {key: self[key]} if key in self else {}
+        super().__delitem__(key)
+        self.notify(removed=removed, added={}, changed={})
+
+    def clear(self):
+        """ Remove all items from the dict. """
+        was_empty = (self == {})
+        removed = self.copy()
+
+        super().clear()
+        if not was_empty:
+            self.notify(removed=removed, added={}, changed={})
+
+    def update(self, other):
+        """ Update the values in the dict by the new dict or an iterable of
+        key-value pairs.
+
+        Parameters
+        ----------
+        other : dict or iterable
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        items = other.items() if isinstance(other, dict) else other
+
+        for key, value in items:
+            validated_key = self.key_validator(key)
+            validated_value = self.value_validator(value)
+
+            if validated_key in self:
+                changed[validated_key] = self[validated_key]
+            else:
+                added[validated_key] = validated_value
+
+            validated_dict[validated_key] = validated_value
+
+        super().update(validated_dict)
+        self.notify(removed={}, added=added, changed=changed)
+
+    def setdefault(self, key, value=None):
+        """ Returns the value if key is present in the dict, else creates the
+        key-value pair and returns the value.
+
+        Parameters
+        ----------
+        key : A hashable object.
+            Key to the item.
+        """
+
+        if key in self:
+            return self[key]
+
+        validated_key = self.key_validator(key)
+        validated_value = self.value_validator(value)
+
+        if validated_key in self:
+            changed = {validated_key: self[validated_key]}
+            added = {}
+        else:
+            changed = {}
+            added = {validated_key: validated_value}
+
+        super().__setitem__(validated_key, validated_value)
+
+        self.notify(removed={}, added=added, changed=changed)
+
+        return validated_value
+
+    def pop(self, key, value=Undefined):
+        """ Remove specified key and return the corresponding
+        value. If key is not found, the default value is returned
+        if given, otherwise KeyError is raised.
+
+        Parameters
+        ----------
+        key : A hashable object.
+            Key to the dict item.
+
+        value : any
+            Value to return if key is absent.
+        """
+
+        should_notify = (value is Undefined or key in self)
+        pop_args = [key, value] if value is not Undefined else [key]
```

Comment:
I think this would be clearer spelled out as:
```
if value is Undefined:
    removed = super().pop(key)
else:
    removed = super().pop(key, value)
```